google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryResultSetImpl.java

```diff
@@ -0,0 +1,554 @@
+
+public class BigQueryResultSetImpl<T> implements BigQueryResultSet<T> {
+
+  private final Schema schema;
+  private final long totalRows;
+  private final BlockingQueue<T> buffer;
+  private T cursor;
+  private final ResultSetWrapper underlyingResultSet;
+  private final BigQueryResultSetStats bigQueryResultSetStats;
+  private final FieldList schemaFieldList;
+
+  public BigQueryResultSetImpl(
+      Schema schema,
+      long totalRows,
+      BlockingQueue<T> buffer,
+      BigQueryResultSetStats bigQueryResultSetStats) {
+    this.schema = schema;
+    this.totalRows = totalRows;
+    this.buffer = buffer;
+    this.underlyingResultSet = new ResultSetWrapper();
+    this.bigQueryResultSetStats = bigQueryResultSetStats;
+    this.schemaFieldList = schema.getFields();
+  }
+
+  @Override
+  public Schema getSchema() {
+    return schema;
+  }
+
+  @Override
+  public long getTotalRows() {
+    return totalRows;
+  }
+
+  @Override
+  public ResultSet getResultSet() {
+    return underlyingResultSet;
+  }
+
+  private class ResultSetWrapper extends AbstractJdbcResultSet {
+    @Override
+    /*Advances the result set to the next row, returning false if no such row exists. Potentially blocking operation*/
+    public boolean next() throws SQLException {
+      try {
+        cursor = buffer.take(); // advance the cursor,Potentially blocking operation
+        if (isEndOfStream(cursor)) { // check for end of stream
+          cursor = null;
+          return false;
+        } else if (cursor instanceof Tuple) {
+          Tuple<Map<String, Object>, Boolean> curTup = (Tuple<Map<String, Object>, Boolean>) cursor;
+          if (!curTup.y()) { // last Tuple
+            cursor = null;
+            return false;
+          }
+        }
+      } catch (InterruptedException e) {
+        throw new SQLException("Error occurred while reading buffer");
+      }
+
+      return true;
+    }
+
+    private boolean isEndOfStream(T cursor) {
+      return cursor instanceof ConnectionImpl.EndOfFieldValueList;
+    }
+
+    @Override
+    public Object getObject(String fieldName) throws SQLException {
+      if (fieldName == null) {
+        throw new SQLException("fieldName can't be null");
+      }
+      if (cursor == null) {
+        return null;
+      } else if (cursor instanceof FieldValueList) {
+        FieldValue fieldValue = ((FieldValueList) cursor).get(fieldName);
+        return fieldValue.getValue() == null ? null : fieldValue.getValue();
+      } else { // Data received from Read API (Arrow)
+        Tuple<Map<String, Object>, Boolean> curTuple = (Tuple<Map<String, Object>, Boolean>) cursor;
+        if (!curTuple.x().containsKey(fieldName)) {
+          throw new SQLException(String.format("Field %s not found", fieldName));
+        }
+        return curTuple.x().get(fieldName);
+      }
+    }
+
+    @Override
+    public Object getObject(int columnIndex) throws SQLException {
+      if (cursor == null) {
```

Comment:
You could reuse most of the code in `getObject(String)` and `getObject(int)` by passing a `Function` lambda for getting a `FieldValue` from a `FieldValueList` -- something like `fvl -> fvl.get(columnIndex)`. Then the logic for closed-result-set / list / read-api can all be reused. In fact, it could be reused across the `getObject()/getString()/getWhatever()` methods, since the overall structure checking for what type of thing cursor is remains the same.

Alternatively, you could use the same approach you have in `getObject(schemaFieldList.get(columnIndex).getName())`, and delegate to a single code path at all times.

Reply:
We will be able to introduce some general pattern but the code will still be repetitive in each method since they all have to be parsed differently. I can see how this increases readability but I think it might also complicate things. Since this is an implementation detail that we don't expect users to interact with, we will defer this to address later.